### DISCUSSION TITLE
Ditch esbuild for TSC

### DISCUSSION THREAD

**Discussion by Doelfke** [remix-run/remix#6456] [Proposals]
While it's faster, it forces Typescript users (the majority of devs) to also run TSC, which negates the performance gains since you essentially need to compile it twice.

Choosing esbuild is not pragmatic or objectively a good idea.  It leads to bugs that can easily be avoided. It makes much more sense to wait an extra 100ms to ensure the code you wrote won't blow up production.

**Comment by sergiodxa** [comment]
Choosing TSC instead of esbuild would mean now the team needs to implement a lot of things that esbuild does that TSC doesn't, like importing non TS files.

I think it's easy enough to add a `npx tsc --noEmit` line in your CI pipeline and a pre-commit hook so you can't commit invalid TS code and if you did you can't merge it or deploy it.

If anything, making `remix dev` does a `tsc` call to check type errors would be simpler and better than ditching esbuild completely.

**Reply by Doelfke** [reply]
It is possible to import javascript files with TSC. 

And running TSC in addition is not better because you are essentially building the project twice, negating the performance benefits of using esbuild.  That’s not a pragmatic approach.

**Reply by sergiodxa** [reply]
> It is possible to import javascript files with TSC

In Remix you can import images, videos, fonts, SQL files, GQL files, and many many more files.

> And running TSC in addition is not better because you are essentially building the project twice, negating the performance benefits of using esbuild. That’s not a pragmatic approach.

You only need to run `tsc` to check types, don't emit any file, and it can run together with the rest of your build. So your app build but you also get type checking to happen.
